File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class TriangleDetection {
private:
    unordered_map<int, unordered_set<int>> adjacency_list;

public:
    void add_edge(int u, int v) {
        adjacency_list[u].insert(v);
        adjacency_list[v].insert(u);
    }
    
    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (const auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }
    
    int count_triangles() {
        int count = 0;
        for (const auto& pair : adjacency_list) {
            int u = pair.first;
            const auto& neighbors = pair.second;
            for (int v : neighbors) {
                if (v > u) {
                    for (int w : neighbors) {
                        if (w > v && adjacency_list[v].find(w) != adjacency_list[v].end()) {
                            count++;
                        }
                    }
                }
            }
        }
        return count;
    }
    
    vector<tuple<int, int, int>> find_triangles() {
        vector<tuple<int, int, int>> triangles;
        for (const auto& pair : adjacency_list) {
            int u = pair.first;
            const auto& neighbors = pair.second;
            for (int v : neighbors) {
                if (v > u) {
                    for (int w : neighbors) {
                        if (w > v && adjacency_list[v].find(w) != adjacency_list[v].end()) {
                            triangles.push_back({u, v, w});
                        }
                    }
                }
            }
        }
        return triangles;
    }
    
    bool has_triangle() {
        for (const auto& pair : adjacency_list) {
            int u = pair.first;
            const auto& neighbors = pair.second;
            for (int v : neighbors) {
                if (v > u) {
                    for (int w : neighbors) {
                        if (w > v && adjacency_list[v].find(w) != adjacency_list[v].end()) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
};

File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.go
package main

type TriangleDetection struct {
    adjacencyList map[int]map[int]bool
}

func NewTriangleDetection() *TriangleDetection {
    return &TriangleDetection{
        adjacencyList: make(map[int]map[int]bool),
    }
}

func (td *TriangleDetection) AddEdge(u, v int) {
    if td.adjacencyList[u] == nil {
        td.adjacencyList[u] = make(map[int]bool)
    }
    if td.adjacencyList[v] == nil {
        td.adjacencyList[v] = make(map[int]bool)
    }
    td.adjacencyList[u][v] = true
    td.adjacencyList[v][u] = true
}

func (td *TriangleDetection) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        td.AddEdge(edge[0], edge[1])
    }
}

func (td *TriangleDetection) CountTriangles() int {
    count := 0
    for u, neighbors := range td.adjacencyList {
        for v := range neighbors {
            if v > u {
                for w := range neighbors {
                    if w > v && td.adjacencyList[v][w] {
                        count++
                    }
                }
            }
        }
    }
    return count
}

func (td *TriangleDetection) FindTriangles() [][3]int {
    triangles := make([][3]int, 0)
    for u, neighbors := range td.adjacencyList {
        for v := range neighbors {
            if v > u {
                for w := range neighbors {
                    if w > v && td.adjacencyList[v][w] {
                        triangles = append(triangles, [3]int{u, v, w})
                    }
                }
            }
        }
    }
    return triangles
}

func (td *TriangleDetection) HasTriangle() bool {
    for u, neighbors := range td.adjacencyList {
        for v := range neighbors {
            if v > u {
                for w := range neighbors {
                    if w > v && td.adjacencyList[v][w] {
                        return true
                    }
                }
            }
        }
    }
    return false
}